### PR TITLE
[WB-7260] Feature python report API

### PR DESCRIPTION
**Pull request by vanpelt**
Fixes WB-7260

Description
-----------
This is a first pass at making a python api for reports, comments welcome.  A basic example looks like:

```python
api = wandb.Api()
# this workflow assumes we're always working from an existing template report
report = api.report("team/username/Report-XXXXX")
# change which runs are selected / visible in a run set
report.run_sets["Run set 2"].visible = ["XXXX", "YYYY"]
# change what filters are applied to a run set
report.run_sets["Run set"].runs = api.runs("vanpelt/foo", filters={"id": {"$in": ["XXXX", "YYYY"]}})
# by default, save will clone the report
report.save("My new report")
# users can also create drafts or overwrite
report.save(draft=True, clone=False)
```

What this isn't doing is allowing users to modify which metrics / visualizations are shown in a given run set.  This might be something we want to expose eventually, but the majority of use-cases can be handled by setting the filters or which runs / groups are currently visible.  Users can also control which run set is currently open and enable / disable them:

```python
report.run_sets["Run set"].open = True
report.run_sets["Run set"].enabled = False
```

> This will only work for v5 reports which means users attempting to modify reports programmatically created more than a few months ago will need to open and save their legacy reports in the UI first.

Testing
-------
How was this PR tested?

It wasn't yet, and it should be 🙃 

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

**Comment by BBischof** [comment]
This is probably unneccessary detail, and probably obvious, but here's what came to mind:

# Use cases

Let's think through a few key use-cases for programmable reports:
- cron-based DS results
- model evaluation pipelines
- automatic leaderboards

From the above I think the above simple API can suffice for all three.

## DS Results

Assume I have my collection of DS models that get retrained and rerun for some weekly inference tasks. Usually I either rerun them by hand, convert them to sql queries for automation, rerun them as automated notebooks (with something like papermill), or launch them as jobs and write the results to some schema for a viz engine to injest.

In the last case, you'd often tag them as most recent and hope the viz engine can handle more sophisticated filtering for knowing which to display. Often you write them to some sql-style schema which the front end app queries, this is how our current repo-insights and orgs pages work (roughly). This also assumes most of the transformations in the front end operate on the schema.

If you're using wandb–a reasonable prior if you have reports already–your model training is already well suited to be displayed in the report format. A new launched run of your data pipeline will now log to wandb as before to generate new runs to consider in your template report. 

This seems very feasible for this pipeline, and it's main comparison would be papermill. With good timeseries support, this could almost completely eliminate our need for something like Mode. (We would have to spend a fair bit of time on improving some chart types to make it look as nice).

## Model eval

In the model evaluation pipeline, like we will soon support via model registry, you really want to have cron-retraining, with some number of pre-built validation test sets and metrics. One of our "competitors", Arize, basically entire product is a convenient interface for scheduling cron jobs that run these validation test sets, and update the performance metrics.

With this structure, you can use the filtering API above to quickly pick the models from the registry you care about (via the runs that train them), and grab their runs data for side-by-side comparison. This will also make seeing things like model drift kinda trivial because you just go to your day/week-ly report and at a glance see the validation runs side-by-side on all your subsamples.

This is especially powerful when trying to evaluate things like group-loss; i.e. pick a covariate and look at the error distribution with that covariate distribution. One thing I haven't seen on our platform that would be sick in this context is using the parallel coords plot to display how varying input parameters translate to validation loss over different run sets. When you can update the runs programmatically, this allows you to automate sensitivity analysis for your top-k models.

## Automatic leaderboards

Kinda iterating on the above, this makes it pretty easy to automate complicated rules for best-k models in a report which serves as a leaderboard. You can imagine anyone able to contribute to some modeling task and at all times the automated report knows the complicated series of evaluations that go into being in the top.

# Opportunities

I don't see exactly how this interacts with artifacts. I know you can log artifacts to your runs, and that might be enough to handle what I'm suggesting above. 

I also wonder about tables. A lot of the power of automatically running new models through a series of inference tasks and automating creation of a report is in the ability to inspect the inference output itself. Being able to at a glance see the difference in labels between the old-best-model and the new-best-model in a pre-created table would be dope. Based on what I've been hearing about model-registry, having part of the promotion flow for new-best-model this auto-generation of a table means that checking up on your model's output is super easy. It means that users will likely update and retrain their model far more often, which I think leads users back to our portal far far more often.

It may be obvious in the above and I'm just ignorant of the current API, but I wonder how convenient it is to filter/slice the run sets. Is it, for example, easy to say: give me all runs in this project with `log_loss < .3` and the `emb_dim < 25`? I kinda implicitly assumed this was easy above.

**Comment by vanpelt** [comment]
Thanks for the brain dump @BBischof!  Creating filters today from python is a little clunky but works.  You need to specify the query using the MongoDB query language and our special prefixes, for example:

```python
run_set.runs = api.runs("wandb/party", filters={"$and": [{"summary_metrics.log_loss": {"$lt": 0.3}}, {"config.emb_dim": {"$lt": 25}]})
```

There are some experiments to make expressing filters more pythonic, we'll end up there one day.  I'm expecting the most common use case for this new api is simply changing what runs are currently driving the visualizations regardless of filters.  This is done by providing a list of run or group ids to a run_set's `visible` attribute.

As far as tables and artifacts are concerned, the current API only allows modifying panels added to a Panel Grid.  A user can add a table to a Panel Grid and they could choose which runs drive it with this api.  Adding the ability to modify weave expressions is something I'm holding off on as the api is still in flux but is coming one day.

**Comment by vanpelt** [comment]
@andrewtruong @vwxyzjn left some great feature requests / users stories we could try to make work in your new branch.

Concerning #2 it's unfortunately pretty tricky.  We would need to run a headless browser somewhere that literally rendered the page and then take a screenshot.  We might be able to support SVG export via a simple backend service, but PNG would be trickier...

**Comment by andrewtruong** [comment]
Hey @vwxyzjn, I think most of what you're looking for is available [here](https://github.com/wandb/client/pull/3732).  Can you try it out to see if it works for you?  Here's an [example colab](https://colab.research.google.com/drive/1os19GzekGqBlwNIjaX8guBydpBXydtM-).

The main gaps I see are exporting images and panel layout.

**Comment by vwxyzjn** [comment]
Hi @andrewtruong, thanks for sharing. The notebook looks really exciting! I have followed the API to create https://wandb.ai/costa-huang/cleanRL/reports/Archived-Atari-CleanRL-s-PPO--VmlldzoyMTAzMjA5 👍

image

It would absolutely solve most use cases for us. 

A couple of nits:

1. if no description is provided to the report, the API seems to refill the description block with "No description". See [code](https://github.com/openrlbenchmark/wandb-rl-plot/blob/d98f48e77aeab838b48e14f9e8599f759c802fa5/plot_rl_games.py#L9-L11)
    Screen Shot 2022-06-01 at 9 46 41 PM
1. multi-level args filtering like `f'params.config.env_config.env_name == "{env_id}" and params.algo.name == "a2c_continuous"'` doesn't work.  See [code](https://github.com/openrlbenchmark/wandb-rl-plot/blob/d98f48e77aeab838b48e14f9e8599f759c802fa5/plot_rl_games.py#L18-L20)

https://user-images.githubusercontent.com/5555347/171530704-31fe7cb5-35ce-4e62-b557-70a1515e78a9.mov


3. there is no way to customize the legend. I usually use the "Advanced Legend" to be `${runsetName}`
4. (minor) no way to customize the color for the lines - in our [documentation](https://docs.cleanrl.dev/rl-algorithms/ppo/#experiment-results_5) we usually use one color to keep uniformity
    image
5. (niche use case) when merging the metrics of different names, there are two issues (see [code](https://github.com/openrlbenchmark/wandb-rl-plot/blob/3037078725cf91a6460ef209419cf5e3beeaa839/plot_merge.py#L32-L53))
    * There is no way to toggle this "grouping - metrics" button, which aggregates over all metrics (in this case it allows us to compare the same metrics under different names from different projects)  
    * The grouping is a bit weird, it produces 4 groups for some reasons, whereas when doing this by hand it only produces 2 groups. See the second video 

https://user-images.githubusercontent.com/5555347/171531936-123f606d-2a63-466f-a585-51b6966cf4eb.mov
https://user-images.githubusercontent.com/5555347/171532845-4f53e0b3-e4db-4f9e-b6b5-56afef92a413.mov

Thanks so much for this awesome PR!

**Comment by andrewtruong** [comment]
Thanks for the feedback @vwxyzjn :)

A few follow-ups:

> 2. multi-level args filtering like `f'params.config.env_config.env_name == "{env_id}" and params.algo.name == "a2c_continuous"'` doesn't work.  See [code](https://github.com/openrlbenchmark/wandb-rl-plot/blob/d98f48e77aeab838b48e14f9e8599f759c802fa5/plot_rl_games.py#L18-L20)

The backend uses dotted names, so we'll need to find a way around this.  Maybe we can do what pandas does with backticks?  Would `f'``params.config.env_config.env_name`` == "{env_id}" and ``params.algo.name`` == "a2c_continuous"'` make sense to you?


>    * The grouping is a bit weird, it produces 4 groups for some reasons, whereas when doing this by hand it only produces 2 groups. See the second video

This might be because you only enabled run set aggregation for the first (but not the second) run set.  Can you try enabling both and seeing if the problem goes away?  In the update we should probably toggle this by default.

**Comment by vwxyzjn** [comment]
I was trying to look for the cause of https://github.com/wandb/client/pull/3202#issuecomment-1144823759 and found this:

image

It looks like the `report.spec` created by hands (right side of the code) puts the `runSets` under `panelBankSectionConfig` compared to the `report.spec`  generated by the API (left side of the code). Maybe this could be the reason that merging metrics under different names behaves weirdly?

**Comment by dmitryduev** [comment]
Closing as stale and likely moot.
